lib/puppet/functions/fact_for.rb

```diff
@@ -0,0 +1,28 @@
+# Copyright (c) 2018 The Regents of the University of Michigan.
+# All Rights Reserved. Licensed according to the terms of the Revised
+# BSD License. See LICENSE.txt for details.
+
+Puppet::Functions.create_function(:fact_for) do
+  dispatch :fact_for do
+    required_param 'String', :node_id
+    required_param 'String', :fact_name
+  end
+
+  def fact_for(node_id, fact_name)
+    fact_keys = fact_name.split('.')
```

Comment:
Lines 12-13 look like one responsibility, 15-20 another, and 22-24 a third. 

22-24 also confuses me: the first item in fact_keys is going to result in overwriting value.  Is there a way to use a non-destructive enumerable operation like `inject` to accomplish this that might be more clear?  (It could also be made clear by having its own method with an appropriate name).

Reply:
There's a lot of ruby features I don't know about, so it wouldn't surprise me to learn that ruby had a shortcut for this. The `fact_keys` array will be empty if there aren't any dots in the fact name, so the loop at the end won't execute at all in that case.

What I'm trying to do is use the first element as the parameter I hand to puppetdb (which doesn't support facter's dot syntax), and then use the rest to select keys; e.g. `one.two.three` becomes `value_for('one')['two']['three']`. Is there a ruby-ism for doing that?

Reply:
```
[:a,:b:c].inject(start){|memo, key| memo.fetch(key) }

# results in:
start.fetch(:a)
start.fetch(:a).fetch(:b)
start.fetch(:a).fetch(:b).fetch(:c)
```
And returns the last one.

Reply:
I ended up using dig instead; looks a bit cleaner to me. Apparently it does pretty much the exact thing I was trying to do.